Fix exponential behavior during error reporting

martint opened the pull request:
Fixes #1407

martint committed:
Process parsing states in LIFO order

Use a stack instead of a queue. This makes it easier to follow
what the error handler is doing during debugging.

martint committed:
Add toString to parsing state to ease debugging

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -148,24 +185,23 @@ public CallerContext(CallerContext parent, ATNState followState)
-            this.atn = atn;
-            this.vocabulary = vocabulary;
+            this.parser = parser;
+            this.stream = parser.getTokenStream();
```
could you also move this assignment last? (stream is a last field)

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -115,14 +115,50 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
+        {
+            Token token = parser.getTokenStream().get(tokenIndex);
+
+            String text = token.getText();
```
nit `firstNonNull(token.getText(), "?")`

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -115,14 +115,50 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
+
+            String text = token.getText();
+            if (text != null) {
+                text = text.replace("\n", "\\n");
```
you may also want (first)
`.replace("\\", "\\\\")`

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -115,14 +115,50 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
+            }
+
+            StringBuilder builder = new StringBuilder();
+            builder.append(String.format("[%s] %s%s:%s @ %s:<%s>:%s",
```
static import,
per code style, add \n before fmt str

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -115,14 +115,50 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
+                text = "?";
+            }
+
+            StringBuilder builder = new StringBuilder();
```
no need for StringBuilder here.

Just
```
String string = format( ... );

if (caller != null) {
    string += format(...);
}

return string;
```

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -144,33 +136,13 @@ public String toString()
 
-            StringBuilder builder = new StringBuilder();
-            builder.append(String.format("[%s] %s%s:%s @ %s:<%s>:%s",
-                    depth,
```
correct me if i'm wrong, this belongs to the previous commit

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -144,33 +136,13 @@ public String toString()
-            StringBuilder builder = new StringBuilder();
-            builder.append(String.format("[%s] %s%s:%s @ %s:<%s>:%s",
-                    depth,
+            return String.format("%s%s:%s @ %s:<%s>:%s",
```
static import,
per code style, add \n before fmt str

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -199,14 +174,59 @@ public Analyzer(
+                    return true;
+                }
+
+                for (int i = 0; i < current.getNumberOfTransitions(); i++) {
```
`for (Transition transition : current.getTransitions()) {`

martint replied:
I chose the current version because `getTransitions()` does an eager copy of the underlying array.

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -199,14 +174,59 @@ public Analyzer(
+            activeStates.add(from);
+
+            while (!activeStates.isEmpty()) {
+                ATNState current = activeStates.pop();
```
is it obvious there cannot be a cycle?

martint replied:
Yes, it's guaranteed there are no cycles via epsilon transitions. Obviously, there could be a cycle via transitions that consume tokens, but those would eventually consume the entire input and finish.

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -199,14 +174,59 @@ public Analyzer(
+            while (!activeStates.isEmpty()) {
+                ATNState current = activeStates.pop();
+
+                if (current.stateNumber == target.stateNumber) {
```
The search could be shorter if you checked the condition before _adding_ to the queue, not after _pop_. (just thinking)

martint replied:
I'd need to duplicate the check in the `add` above the loop (in case target == from) and before the `push()` below. I thought it unnecessarily complicated the code for very little practical gain.

findepi commented on presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java:
```diff
@@ -112,7 +112,7 @@
                         "line 1:36: mismatched input '<EOF>'. Expecting: <expression>"},
                 {"SELECT a AS z FROM t WHERE a. ",
-                        "line 1:29: mismatched input '.'. Expecting: '%', '*', '+', '-', '/', 'AT', '||', <expression>"},
+                        "line 1:29: mismatched input '.'. Expecting: '%', '*', '+', '-', '/', 'AND', 'AT', 'EXCEPT', 'FETCH', 'GROUP', 'HAVING', 'INTERSECT', 'LIMIT', 'OFFSET', 'OR', 'ORDER', 'UNION', '||', <EOF>, <predicate>"},
```
I wonder why the `.` is the problem. It could be saying:
```
unexpected EOF, expecting: <identifier> (field)
```

martint replied:
That's a good question. Let me investigate.

martint replied:
It's because Antlr bails out and leaves the the parser at a state past `>primaryExpression>`, which is the one that recognizes `<identifier> . <identifier>`. I guess it's matching `<identifier>` greedily as a `<primaryExpression>' because that's all that it can match given the available tokens in the stream and then it says "oops, what do I do with this `.`" now?

findepi commented on presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java:
```diff
@@ -139,6 +139,23 @@
         };
     }
 
+    @Test
```
add timeout

findepi commented on presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java:
```diff
@@ -139,6 +139,23 @@
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9",
+                "line 1:375: mismatched input '<EOF>'. Expecting: '%', '*', '+', '-', '/', 'AT', 'THEN', '||'");
```
i'd expect `IS` too, for example in `CASE WHEN {multiplication} IS NULL THEN 'zonk' ELSE ..`

(just noting, no change requested)

martint replied:
That's partly because our grammar does: `CASE <valueExpression> WHEN ...`. Looking at the spec, it seems to allow more complex expressions, including our `<booleanExpression>` (I'll fix that as a separate change). Even them, it suffers from the same problem I explained regarding `SELECT ... FROM ... WHERE a.`

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -345,4 +376,26 @@ public ErrorHandler build()
+
+    private static class Result
+    {
+        private final int errorToken;
```
`Token`? or name it `errorTokenIndex`

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -91,17 +90,14 @@ public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int
                     .collect(Collectors.joining(", "));
 
-            message = format("mismatched input '%s'. Expecting: %s", ((Token) offendingSymbol).getText(), expected);
+            message = format("mismatched input '%s'. Expecting: %s", parser.getTokenStream().get(result.getErrorToken()).getText(), expected);
```
Using `result.getErrorToken()` (`furthestToken`) there is a theoretical possibility we will say "mismatched" at quoting something else than ANTLR choked at. 

I replaced `parser.getTokenStream().get(result.getErrorToken())` with `((Token) offendingSymbol)` and it didnt have any noticable impact.
Maybe at least something like
```
if (((Token) offendingSymbol).getTokenIndex() != result.getErrorToken()) {
   log.warn ..
}
```
?

martint replied:
It doesn't really matter. All we care about is that there is *some* syntax error and we want to produce a reasonable error. There's no false positive issue here where we might report an error where there is none.

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -273,41 +293,52 @@ else if (transition instanceof WildcardTransition) {
 
-                        if (labels.contains(currentToken)) {
-                            activeStates.push(new ParsingState(transition.target, tokenIndex + 1, false, caller, parser, current.depth + 1));
+                        if (labels.contains(currentToken) && tokenIndex < stream.size() - 1) {
```
is `&& tokenIndex < stream.size() - 1` necessary?
i think EOF token serves as a sentinel.

martint replied:
I had to add this in one of my earlier experiments, but you're right. Let me see if I can change it back.

findepi commented on presto-parser/src/main/java/io/prestosql/sql/parser/ErrorHandler.java:
```diff
@@ -273,41 +293,52 @@ else if (transition instanceof WildcardTransition) {
+                }
+
+                candidates.addAll(labels);
+                furthestToken = tokenIndex;
```
move this assignment inside `if (tokenIndex > furthestToken) {`

findepi reviewed (approved):
Couple of minor comments

martint committed:
Consider rule precedence during error handling

The ATN simulator was not considering rule precendence
markers and transitions, which resulted in rules being
evaluated an exponential number of times for a query like:

  SELECT CASE 1 * 2 * 3 * 4 * 5
